pkg/repo/mysql/translation.go

```diff
@@ -14,94 +14,85 @@ import (
-	var translationID uuid.UUID
-
-	row := tx.QueryRowContext(
-		ctx,
-		`SELECT id FROM translation WHERE service_id = UUID_TO_BIN(?) AND language = ?`,
-		serviceID,
-		translation.Language.String(),
-	)
-
-	// Check if translation already exists, if not, create a new one
-	switch err = row.Scan(&translationID); {
-	// Translation already exists
-	default:
-		// noop
+	return r.ensureTx(ctx, func(ctx context.Context, r *Repo) error {
+		_, err := r.LoadService(ctx, serviceID)
+		if err != nil {
+			return fmt.Errorf("repo: load service: %w", err)
+		}
 
-	// Translation does not exist
-	case errors.Is(err, sql.ErrNoRows):
-		translationID = uuid.New()
+		// Check if translation already exist
+		var translationID uuid.UUID
 
-		if _, err = tx.ExecContext(
+		row := r.db.QueryRowContext(
 			ctx,
-			`INSERT INTO translation (id, service_id, language, original) VALUES (UUID_TO_BIN(?), UUID_TO_BIN(?), ?, ?)`,
-			translationID,
+			`SELECT id FROM translation WHERE service_id = UUID_TO_BIN(?) AND language = ?`,
 			serviceID,
 			translation.Language.String(),
-			translation.Original,
-		); err != nil {
-			return fmt.Errorf("repo: insert message: %w", err)
-		}
+		)
 
-	// Error scanning row
-	case err != nil:
-		return fmt.Errorf("repo: scan message: %w", err)
-	}
+		// Check if translation already exists, if not, create a new one
+		switch err = row.Scan(&translationID); {
+		// Translation already exists
+		default:
+			// noop
+
+		// Translation does not exist
+		case errors.Is(err, sql.ErrNoRows):
+			translationID = uuid.New()
+
+			if _, err = r.db.ExecContext(
+				ctx,
+				`INSERT INTO translation (id, service_id, language, original) VALUES (UUID_TO_BIN(?), UUID_TO_BIN(?), ?, ?)`,
+				translationID,
+				serviceID,
+				translation.Language.String(),
+				translation.Original,
+			); err != nil {
+				return fmt.Errorf("repo: insert message: %w", err)
+			}
 
-	// Insert into message table,
-	// on duplicate message.id and message.translation_id,
-	// update message's message, description and status values.
-	stmt, err := tx.PrepareContext(
-		ctx,
-		`INSERT INTO message
-	(translation_id, id, message, description, positions, status)
-VALUES
-	(UUID_TO_BIN(?), ?, ?, ?, ?, ?)
-ON DUPLICATE KEY UPDATE
-	message = VALUES(message),
-	description = VALUES(description),
-	positions = VALUES(positions),
-	status = VALUES(status)`,
-	)
-	if err != nil {
-		return fmt.Errorf("repo: prepare stmt to insert message: %w", err)
-	}
-	defer stmt.Close()
+		// Error scanning row
+		case err != nil:
+			return fmt.Errorf("repo: scan message: %w", err)
+		}
 
-	for _, m := range translation.Messages {
-		_, err = stmt.ExecContext(
+		// Insert into message table,
+		// on duplicate message.id and message.translation_id,
+		// update message's message, description and status values.
+		stmt, err := r.db.PrepareContext(
 			ctx,
-			translationID,
-			m.ID,
-			m.Message,
-			m.Description,
-			m.Positions,
-			m.Status,
+			`INSERT INTO message
```

Comment:
Please ensure the target consumer of this SQL statement is an observability user. This applies to ALL projects we work on.

Not a developer.